### PR TITLE
Afform - Implement client-side validation of required fields

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
This is (so far) a client-side only implementation of required field validation for Afforms.

See https://lab.civicrm.org/dev/core/-/issues/3137

Before
----------------------------------------
Required fields not required.

After
----------------------------------------
image

### CODE REVIEW

File: ext/afform/core/ang/af/afForm.component.js [hunk 1]
```diff
@@ -92,7 +99,7 @@
           values: data}
         ).then(function(response) {
           if (ctrl.fileUploader.getNotUploadedItems().length) {
-            _.each(ctrl.fileUploader.getNotUploadedItems(), function(file) {
+            _.each(ctrl.fileUploader.getNotUploadedItegitk&ms(), function(file) {
```
Comment (MegaphoneJon) [inline ext/afform/core/ang/af/afForm.component.js:102]: This looks unintentional.

Comment (colemanw) [reply ext/afform/core/ang/af/afForm.component.js]: Oopa

**Review by kurund — CHANGES_REQUESTED** [review]
@colemanw 

Required validation seems to be working fine on backend and frontend for standard fields.

It needs to be fixed for block like Email block.

**Comment by aydun** [comment]
It's working for fields from the standard entities but not for fields from my custom entity.  The fields show as required on the builder page but not on the display page.  

To test with a custom entity, see https://lab.civicrm.org/dev/core/-/issues/3728

**Comment by eileenmcnaughton** [comment]
@colemanw this needs a rebase - also not quite sure status otherwise but can get @kurund to recheck once it is not stale

**Comment by colemanw** [comment]
I've rebased this @eileenmcnaughton @kurund

**Comment by kurund** [comment]
@colemanw 

Below is the sample form with all the fields configured as required in the Formbuilder.
Screenshot from 2022-08-01 10-48-16

However, I was able to submit the form by filling only data in the above screenshot.

**Comment by kurund** [comment]
@colemanw 

It looks like "required" param is missing in field definition for 'joins'. You can check the generated code. Hence, I the email validation is failing.

**Comment by colemanw** [comment]
@kurund I've fixed setting default required value for e.g. Email joins.
I think there's still some issue with Select2, but maybe this is good enough to merge now and we can investigate Select2 in a different PR.

**Review by kurund — APPROVED** [review]
Looks good to me. ready to for merging.

TODO for separate PR
- Select2 fields
- RichText fields

**Pull request merged**